File: src/main/java/gargoyle/netsquares/util/Assert.java
package gargoyle.netsquares.util;

import java.text.MessageFormat;
import java.util.Objects;
import java.util.concurrent.Callable;

public final class Assert {
    private Assert() {
    }

    public static void assertEmpty(String value, String message, Object... args) {
        check(() -> value == null || value.trim().isEmpty(), message, args);
    }

    private static void check(Callable<Boolean> operation, String message, Object... args) {
        try {
            if (operation.call()) {
                return;
            }
        } catch (Exception e) {
            //
        }
        throw new IllegalArgumentException(MessageFormat.format(message, args));
    }

    public static <T> void assertEquals(T actual, T expected, String message, Object... args) {
        check(() -> Objects.equals(actual, expected), message, args);
    }

    public static <T extends Comparable<T>> void assertInRange(T actual, T min, T max, String message, Object... args) {
        check(() -> (min == null || min.compareTo(actual) <= 0) && (max == null || max.compareTo(actual) >= 0), message, args);
    }

    public static void assertNotEmpty(String value, String message, Object... args) {
        check(() -> value != null && !value.trim().isEmpty(), message, args);
    }

    public static <T> void assertNotEquals(T actual, T expected, String message, Object... args) {
        check(() -> !Objects.equals(actual, expected), message, args);
    }

    public static <T extends Comparable<T>> void assertNotInRange(T actual, T min, T max, String message, Object... args) {
        check(() -> min != null && min.compareTo(actual) > 0 && (max == null || max.compareTo(actual) >= 0), message, args);
    }

    public static <T> void assertNotNull(T value, String message, Object... args) {
        check(() -> value != null, message, args);
    }

    public static <T> void assertNotSame(T actual, T expected, String message, Object... args) {
        check(() -> actual != expected, message, args);
    }

    public static <T> void assertSame(T actual, T expected, String message, Object... args) {
        check(() -> actual == expected, message, args);
    }

    public static void assertTrue(boolean value, String message, Object... args) {
        check(() -> value, message, args);
    }
}

File: src/main/kotlin/gargoyle/netsquares/util/Assert.kt
package gargoyle.netsquares.util

object Assert {
    fun assertEmpty(value: String?, message: () -> String) {
        require(value == null || value.trim().isEmpty(), message)
    }

    fun <T : Any> assertEquals(actual: T, expected: T, message: () -> String) {
        require(actual == expected, message)
    }

    fun <T : Comparable<T>> assertInRange(actual: T, min: T, max: T, message: () -> String) {
        require((min <= actual) && (max >= actual), message)
    }

    fun assertNotEmpty(value: String?, message: () -> String) {
        require(value != null && value.trim().isNotEmpty(), message)
    }

    fun <T : Any> assertNotEquals(actual: T, expected: T, message: () -> String) {
        require(actual != expected, message)
    }

    fun <T : Comparable<T>> assertNotInRange(actual: T, min: T, max: T, message: () -> String) {
        require(min > actual && (max >= actual), message)
    }

    fun <T : Any> assertNotSame(actual: T, expected: T, message: () -> String) {
        require(actual !== expected, message)
    }

    fun <T : Any> assertSame(actual: T, expected: T, message: () -> String) {
        require(actual === expected, message)
    }

    fun assertTrue(value: Boolean, message: () -> String) {
        require(value, message)
    }
}
